File: haber-api/src/models/news.js
const mongoose = require("mongoose")
const RSS = require("../models/rss")
const NewsSite = require("../models/news_site")

const newsSchema = new mongoose.Schema({
    rss: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'RSS',
        required: true,
    },
    link: {
        type: String,
        required: true,
        unique: true
    },
    title: {
        type: String,
        text: true,
        required: true,
    },
    description: {
        type: String,
        text: true,
        required: true,
    },
    body: {
        type: String,
        required: true,
        text: true,
    },
    date: {
        type: Number,
        required: true,
    },
    image: {
        type: String,
        required: true,
    },
    likes: [{
        users: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User'
        }
    }],
    dislikes: [{
        users: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User'
        }
    }],
    viewers: {
        type: Number,
        default : 0,
    },
    viewers_unique: [{
        firebaseID: {
            type: String,
        },
        count : {
            type: Number,
            default : 0,
        },
    }],
})

newsSchema.methods.toJSON = async function () {
    const news = this
    const newsObject = news.toObject()

    delete newsObject.__v
    delete newsObject.rss

    const rss = await RSS.findOne({
        _id: news.rss
    })

    if (!rss) {
        console.log("RSS not found.")
        return undefined
    }

    const newsSite = await NewsSite.findOne({
        _id: rss.site
    })

    if (!rss) {
        console.log("News Site not found.")
        return undefined
    }


    newsSiteName = newsSite.name
    newsCategoryName = rss.category


    const object = {
        ...newsObject,
        newsSiteName,
        newsCategoryName
    }

    return object
}

const News = mongoose.model("News", newsSchema)

module.exports = News

File: haber/lib/models/News.dart
import 'package:haber/models/Category.dart';
import 'package:haber/models/News_Site.dart';
import 'package:haber/models/RSS.dart';

class News {
  String sId;
  int viewers;
  String title;
  String description;
  String body;
  String date;
  String link;
  String image;
  RssDetails rssDetails;
  SiteDetails siteDetails;
  CategoryDetails categoryDetails;
  int likes;
  int dislikes;
  bool isLiked;
  bool isDisliked;
  bool isFavorited;
  int uniqueViews;

  News({
    this.sId,
    this.viewers,
    this.title,
    this.description,
    this.body,
    this.date,
    this.link,
    this.image,
    this.rssDetails,
    this.siteDetails,
    this.categoryDetails,
    this.likes,
    this.dislikes,
    this.isLiked,
    this.isDisliked,
    this.isFavorited,
    this.uniqueViews,
  });

  static List<String> coded = [
    "&#39;",
    "&#40;",
    "&#41;",
    "&#44;",
    "&#58;",
    "&#59;"
  ];
  static List<String> decoded = ["'", "(", ")", ",", ":", ";"];
  Map<String, String> map = new Map.fromIterables(coded, decoded);

  News.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    viewers = json['viewers'];
    title = json['title'];
    description = json['description'];

    body = json['body'];
    date = json['date'];
    link = json['link'];
    image = json['image'];
    isLiked = json['isLiked'];
    isDisliked = json['isDisliked'];
    isFavorited = json['isFavorited'];

    if (title != null) {
      title =
          map.entries.fold(title, (prev, e) => prev.replaceAll(e.key, e.value));
    }

    if (body != null) {
      body =
          map.entries.fold(body, (prev, e) => prev.replaceAll(e.key, e.value));
    }
    if (description != null) {
      description = map.entries
          .fold(description, (prev, e) => prev.replaceAll(e.key, e.value));
    }

    rssDetails = json['rssDetails'] != null
        ? new RssDetails.fromJson(json['rssDetails'])
        : null;
    siteDetails = json['siteDetails'] != null
        ? new SiteDetails.fromJson(json['siteDetails'])
        : null;
    categoryDetails = json['categoryDetails'] != null
        ? new CategoryDetails.fromJson(json['categoryDetails'])
        : null;
    likes = json['likes'];
    dislikes = json['dislikes'];
    uniqueViews = json['uniqueViews'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['viewers'] = this.viewers;
    data['title'] = this.title;
    data['description'] = this.description;
    data['body'] = this.body;
    data['date'] = this.date;
    data['link'] = this.link;
    data['image'] = this.image;
    if (this.rssDetails != null) {
      data['rssDetails'] = this.rssDetails.toJson();
    }
    if (this.siteDetails != null) {
      data['siteDetails'] = this.siteDetails.toJson();
    }
    if (this.categoryDetails != null) {
      data['categoryDetails'] = this.categoryDetails.toJson();
    }
    data['likes'] = this.likes;
    data['dislikes'] = this.dislikes;

    data['is_liked'] = this.isLiked;
    data['isDisliked'] = this.isDisliked;
    data['isFavorited'] = this.isFavorited;
    data['uniqueViews'] = this.uniqueViews;

    return data;
  }
}
